logging/journald_core.go

```diff
@@ -74,11 +72,55 @@ func (c *journaldCore) Write(ent zapcore.Entry, fields []zapcore.Field) error {
 	return journald.Send(message, pri, enc.Fields)
 }
 
+// addFields adds all given fields to enc with an altered key, prefixed with the journaldCore.identifier and sanitized
+// via journaldFieldEncode.
 func (c *journaldCore) addFields(enc zapcore.ObjectEncoder, fields []zapcore.Field) {
 	for _, field := range fields {
-		field.Key = c.identifierU +
-			"_" +
-			strcase.ScreamingSnake(field.Key)
+		field.Key = journaldFieldEncode(c.identifier + "_" + field.Key)
 		field.AddTo(enc)
 	}
 }
+
+// journaldFieldEncode alters an input string to be used as a journald field key.
+//
+// When journald receives a field with an invalid key, it silently discards this field. This makes syntactically correct
+// keys a necessity. Unfortunately, there was no specific documentation about the field key syntax available. This
+// function follows the logic enforced in systemd's journal_field_valid function[0].
+//
+// This boils down to:
+// - Key length MUST be within (0, 64] characters.
+// - Key MUST start with [A-Z].
+// - Key characters MUST be [A-Z0-9_].
+//
+//	[0]: https://github.com/systemd/systemd/blob/11d5e2b5fbf9f6bfa5763fd45b56829ad4f0777f/src/libsystemd/sd-journal/journal-file.c#L1703
+func journaldFieldEncode(key string) string {
+	if len(key) == 0 {
+		// While this is definitely an error, panicking would be too destructive and silently dropping fields is against
+		// the very idea of ensuring key conformity.
+		return "EMPTY_KEY"
+	}
+
+	isAlpha := func(r rune) bool { return 'A' <= r && r <= 'Z' }
+	isDigit := func(r rune) bool { return '0' <= r && r <= '9' }
+
+	keyParts := []rune(strcase.ScreamingSnake(key))
+	for i, r := range keyParts {
+		if isAlpha(r) || isDigit(r) || r == '_' {
+			continue
+		}
+		keyParts[i] = '_'
+	}
+	key = string(keyParts)
+
+	if !isAlpha(rune(key[0])) {
```

Comment:
If I'm not mistaken, Go already provides utilities for this:

```suggestion
	key := strcase.ScreamingSnake(key)
	for i, r := range key {
		if !(unicode.IsUpper(r) || unicode.IsDigit(r) || r == '_') {
			continue
		}

		key[i] = '_'
	}

	if !unicode.IsUpper(rune(key[0])) {
```

Reply:
journald requires ASCII A-Z, whereas (as the package name already suggests), `unicode.IsUpper()` also returns true for non-ASCII uppercase characters like `Ä`: https://go.dev/play/p/ayqkoWoCL8E

`isUpper` would still be a better name than `isAlpha` though.

Reply:
Setting `key[i] = '_'` in a string is not supported. To do so, it needs to be converted to a `[]rune` or some shenanigans like `key[:i] + "_" + key[i+1:]` are necessary. However, the second option messes up some Unicode characters being bigger than a byte. For example: <https://go.dev/play/p/qo3nfS83ybc>

Reply:
> journald requires ASCII A-Z, whereas (as the package name already suggests), `unicode.IsUpper()` also returns true for non-ASCII uppercase characters like `Ä`

My bad, I didn't expect UTF-8 strings to be supported and overlooked the test for it.